Algorithm/QCAlgorithm.Python.cs

```diff
@@ -317,46 +317,46 @@ public void RegisterIndicator(Symbol symbol, PyObject indicator, TimeSpan? resol
         /// <param name="selector">Selects a value from the BaseData send into the indicator, if null defaults to a cast (x => (T)x)</param>
         public void RegisterIndicator(Symbol symbol, PyObject indicator, IDataConsolidator consolidator, PyObject selector = null)
         {
-            object managedObject = null;
-
             using (Py.GIL())
             {
-                var pythonType = indicator.GetPythonType();
-                if (pythonType.Repr().Contains("QuantConnect"))
-                {
-                    managedObject = indicator.AsManagedObject(pythonType.As<Type>());
-                }
-                else if (!indicator.HasAttr("Update"))
-                {
-                    throw new ArgumentException($"Update method must be defined. Please checkout {indicator}");
-                }
-            }
-
-            // Lean indicators are directed to other RegisterIndicator overloads
-            if (managedObject != null)
-            {
-                var indicatorDataPoint = managedObject as Indicator;
-                if (indicatorDataPoint != null)
+                try
                 {
-                    var managedSelector = (Func<IBaseData, decimal>)selector?.AsManagedObject(typeof(Func<IBaseData, decimal>));
-                    RegisterIndicator(symbol, indicatorDataPoint, consolidator, managedSelector);
-                }
+                    // QuantConnect Indicators
+                    var indicatorPythonType = indicator.GetPythonType();
+                    var indicatorType = GetIndicatorBaseType(indicatorPythonType.As<Type>());
+                    var genericArgumentType = indicatorType.GetGenericArguments().FirstOrDefault();
 
-                var indicatorDataBar = managedObject as BarIndicator;
-                if (indicatorDataBar != null)
-                {
-                    var managedSelector = (Func<IBaseData, IBaseDataBar>)selector?.AsManagedObject(typeof(Func<IBaseData, IBaseDataBar>));
-                    RegisterIndicator(symbol, indicatorDataBar, consolidator, managedSelector);
+                    if (genericArgumentType == typeof(IndicatorDataPoint))
+                    {
+                        var managedIndicator = indicator.AsManagedObject(indicatorType) as IndicatorBase<IndicatorDataPoint>;
+                        var managedSelector = PythonUtil.ToFunc<IBaseData, decimal>(selector);
+                        RegisterIndicator(symbol, managedIndicator, consolidator, managedSelector);
+                    }
+                    else if (genericArgumentType == typeof(IBaseDataBar))
+                    {
+                        var managedIndicator = indicator.AsManagedObject(indicatorType) as IndicatorBase<IBaseDataBar>;
+                        var managedSelector = PythonUtil.ToFunc<IBaseData, IBaseDataBar>(selector);
+                        RegisterIndicator(symbol, managedIndicator, consolidator, managedSelector);
+                    }
```

Comment:
Can we extract this code into a class and put that class under unit test? Any time we get into the weeds with reflection code it's very important to isolate the code. I'm not recommending we remove the integration test you've defined as it confirms it works properly from the algorithm's point of view, but I suspect we'll have issues w/ this code. For example, we talk about `TradeBar` indicators but don't support `QuoteBar`

Reply:
Agree. I am going to create the helper class with unit tests and make a separate PR. After it is reviewed and merged, I will rebase this one.